addons/binding/org.openhab.binding.osramlightify/src/main/java/org/openhab/binding/osramlightify/internal/messages/LightifyListGroupsMessage.java

```diff
@@ -0,0 +1,164 @@
+import org.openhab.binding.osramlightify.internal.exceptions.LightifyMessageTooLongException;
+import org.openhab.binding.osramlightify.internal.messages.LightifyBaseMessage;
+
+/**
+ * Get the firmware version of a Lightify gateway.
+ *
+ * @author Mike Jagdis - Initial contribution
+ */
+public final class LightifyListGroupsMessage extends LightifyBaseMessage implements LightifyMessage {
+
+    private byte[] deviceName = new byte[NAME_LENGTH];
+    private ByteBuffer encodedMessage;
+
+    public LightifyListGroupsMessage() {
+        super(null, Command.LIST_GROUPS);
+
+        try {
+            encodedMessage = super.encodeMessage(1)
+                .put((byte) 0x01);
+        } catch (LightifyMessageTooLongException e) {
+        }
+    }
+
+    @Override
+    public String toString() {
+        return super.toString();
+    }
+
+    @Override
+    public boolean isPoller() {
+        return true;
+    }
+
+    // ****************************************
+    //      Request transmission section
+    // ****************************************
+
+    @Override
+    public ByteBuffer encodeMessage() throws LightifyMessageTooLongException {
+        return super.encodeMessage(encodedMessage);
+    }
+
+    // ****************************************
+    //        Response handling section
+    // ****************************************
+
+    private String makeGroupAddress(short id) {
+        return String.format("00:00:00:00:00:00:%02X:%02X",
+            ((id >> 8) & 0xff), (id & 0xff));
+    }
+
+    @Override
+    public boolean handleResponse(LightifyBridgeHandler bridgeHandler, ByteBuffer data) throws LightifyException {
+        super.handleResponse(bridgeHandler, data);
+
+        short deviceCount = data.getShort();
+
+        ThingUID bridgeUID = bridgeHandler.getThing().getUID();
+
+        HashMap<Short, Object> known = bridgeHandler.knownGroups;
+        @SuppressWarnings("unchecked")
+        HashMap<Short, Object> toRemove = (HashMap<Short, Object>) known.clone();
+
+        for (int i = 0; i < deviceCount; i++) {
+            short deviceNumber = data.getShort();
+            toRemove.remove(deviceNumber);
+
+            data.get(deviceName);
+
+            Object obj = known.get(deviceNumber);
+
+            LightifyDeviceHandler deviceHandler = null;
+            ThingUID thingUID = null;
+
+            if (obj != null && obj instanceof LightifyDeviceHandler) {
+                deviceHandler = (LightifyDeviceHandler) obj;
+
+                // If we have a thing but it is no longer initialized we put it straight back in
+                // the inbox and carry on.
+                if (!deviceHandler.isStatusInitialized()) {
+                    Thing thing = deviceHandler.getThing();
+
+                    // N.B. The IEEE address can't have changed but the name on the gateway might have.
+                    bridgeHandler.getDiscoveryService().discoveryResult(thing.getUID(), THING_TYPE_LIGHTIFY_GROUP,
```

Comment:
The gateway tells you about paired devices, there's no other way to figure them out, therefore the gateway handler has to create additional things (the paired devices), especially since this is the only way to find their addresses and device types.

Reply:
Why does the handler need to do that? I understand that you have to have a connection to the gateway to obtain the list of paired devices. So it is fine to send this message from within your discovery service, but not from your handler.

In the [Philips Hie binding](https://github.com/eclipse/smarthome/tree/master/extensions/binding/org.eclipse.smarthome.binding.hue) the discovery service has to find the bridge first too. Afterwards it can obtain a list of paired devices. There is no need for a handler to create discovery results.